web3c/oasis.js

```diff
@@ -154,9 +143,78 @@ class Oasis {
     this.accounts = accounts;
   }
 
+  _setupSubscribe(web3) {
+    this.subscribe = (subscription, options, callback) => {
+      // Only try to decrypt logs.
+      if (subscription !== 'logs') {
+        return web3.eth.subscribe(subscription, options, callback);
+      }
+
+      let wrappedEmitter = new EventEmitter();
+
+      let decryptSubscriptionLog = async (msg, log) => {
+        try {
+          if (await this.isConfidential(log.address)) {
+            let transform = new ConfidentialSendTransform(null, this.keyManager);
+            await transform.tryDecryptLogs([log], false);
+            wrappedEmitter.emit(msg, log);
+          } else {
+            wrappedEmitter.emit(msg, log);
+          }
+        } catch (err) {
+          wrappedEmitter.emit('error', err);
+        }
+      };
+
+      let wrappedCallback = undefined;
+      if (callback) {
+        wrappedCallback = async (err, log) => {
+          if (err) {
+            return callback(err);
+          }
+          if (await this.isConfidential(log.address)) {
+            let transform = new ConfidentialSendTransform(null, this.keyManager);
+            await transform.tryDecryptLogs([log], false);
+            return callback(null, log);
+          } else {
+            return callback(null, log);
+          }
+        };
+      }
+
+      web3
+        .eth
+        .subscribe(subscription, options, wrappedCallback)
+        .on('data', async (log) => {
+          await decryptSubscriptionLog('data', log);
+        }).on('changed', async (log) => {
+          await decryptSubscriptionLog('changed', log);
+        }).on('error', (err) => {
+          wrappedEmitter.emit('error', err);
+        });
+
+      return wrappedEmitter;
+    };
+  }
+
   resetKeymanager() {
     this.keyManager.reset();
   }
+
+  /**
+   * @returns true iff the contract at address is confidential.
+   */
+  isConfidential(address) {
+    return new Promise((resolve, reject) => {
```

Comment:
likewise, can this promise be replaced with an async function?

Reply:
I believe this one can't because `keyManager.get` takes a callback and I want `isConfidential` to not take a callback (hence the promise).